### ISSUE TITLE
Vehicle physics pro (VPP) vehicle import?

### ISSUE THREAD

**Issue by JeyP4** [lgsvl/simulator#1959]
Hello

I have been through the below posts
[LGSVL-VPP (Vehicle Physics Pro) Integration #515
](https://github.com/lgsvl/simulator/issues/515)&
[Post#1113
](https://github.com/lgsvl/simulator/issues/1113#issuecomment-740113330)
But they couldn't lead to a deterministic approach.

For good autonomous driving simulation, detailed vehicle dynamics is better to have.
Those posts are also old. I was wondering if there is any more-updated approach is available to import VPP vehicles in LGSVL.

Could anyone suggest something regarding it?

**Comment by EricBoiseLGSVL** [comment]
Are you looking for the center of mass?

**Comment by JeyP4** [comment]
> Are you looking for the center of mass?

No, I am looking for rotational Inertia which is also known as moment of inertia kg-m2.

This is missing in Jaguar prefab. This is a critical parameter in real vehicle dynamics.

**Comment by EricBoiseLGSVL** [comment]
So [this](https://docs.unity3d.com/ScriptReference/Rigidbody-inertiaTensor.html) and/or [this](https://docs.unity3d.com/ScriptReference/Rigidbody-inertiaTensorRotation.html)?

**Comment by JeyP4** [comment]
> So [this](https://docs.unity3d.com/ScriptReference/Rigidbody-inertiaTensor.html) and/or [this](https://docs.unity3d.com/ScriptReference/Rigidbody-inertiaTensorRotation.html)?

Thanks for replying.
Yes, I am aware of rigid body tensor. But they are automatically calculated by the mesh collider. And mesh collider assumes uniform weight distribution (which is an oversimplification). That's why in the single-track vehicle model,  a reasonable `J` is used.

![image](https://user-images.githubusercontent.com/44414900/150222447-3fd44c85-ed5d-45d8-bb23-3c31b8a02f90.png)

For example, in [Vehicle physics pro](https://vehiclephysics.com/blocks/inertia/#explicit) correct J value is 2200 kg-m2. Rigid body collider calculates ~3100 kg-m2.
![image](https://user-images.githubusercontent.com/44414900/150222832-a55f6238-7741-4625-9255-0be2d9f1992e.png)

VPP contains a very nice vehicle prefab (In the community version itself), it would be an awesome value addition to LGSVL if a connecting component is available. Maybe for future expansion. As VPP is a good option for realistic vehicle dynamics simulation. 😀

**Comment by EricBoiseLGSVL** [comment]
Ah, I see. I wasn't aware it allowed you to set this. Yes, VPP is great.  We have integrated VPP but we can't release because of license issues. Some other users have as well.

We implemented custom dynamics and VPP can be added to the binary.  What is the issue that you are having integrating exactly?

**Comment by JeyP4** [comment]
> What is the issue that you are having integrating exactly?

I have converted VPP material to HDRP. So now it is ready to be seen in LGSVL scene in unity.
Then I dragged Jaguar and VPP vehicle in a scene to compare them.
![image](https://user-images.githubusercontent.com/44414900/150317459-e7306936-9fdb-4145-afcb-b30a5e134134.png)

I have seen the [LGSVL - Creating a new ego vehicle tutorial](https://www.svlsimulator.com/docs/simulation-content/add-new-ego-vehicle/). I think this tutorial is valid in case of just modifying the vehicle dimensions and look, keeping the same vehicle dynamics.

I have seen the [LGSVL - EGO Vehicle Dynamics](https://www.svlsimulator.com/docs/simulation-content/ego-vehicle-dynamics/#ego-vehicle-dynamics). But I don't know which section is useful for me 'Full Model Interface' or 'Robotics'.

Basically, I am confused which are the next steps for me.

**Comment by EricBoiseLGSVL** [comment]
Like the post you linked earlier, you need to create a new dynamicsSMI class that works with VPP.

**Comment by lemketron** [comment]
@JeyP4 
> Basically, I am confused which are the next steps for me.

You can check out my VehicleVPP.cs file in a VPP fork of SVL that includes the (free) VPP community DLL here:
https://github.com/lemketron/simulator/tree/WIP-AddVPP-on-2021.3

This work was used by some of the [Indy Autonomous Challenge](https://www.indyautonomouschallenge.com/) teams for simulation prior to their autonomous racing events in Indianapolis last October and at CES in Las Vegas in January, 2022.

Hope this helps.

It would be great to have an open source vehicle with the VPP changes in it, but unfortunately we don't have anything available that we can share publicly share.

**Comment by EricBoiseLGSVL** [comment]
Thanks @lemketron

**Comment by JeyP4** [comment]
Thanks @lemketron

I tried using the scripts you suggested in the following way.
![image](https://user-images.githubusercontent.com/44414900/150654381-c7d91001-d696-437b-97d9-77424f6bf704.png)

So some scripts from VPP default, some from your repo, and some from Jaguar. Is it the correct way?
But in simulation, the vehicle is sinking inside the ground.
![Kazam_screencast_00006](https://user-images.githubusercontent.com/44414900/150654710-701d00ba-b659-44c8-b98f-782cabdf1559.gif)
I think it stops when body 'Collider' hits the ground.

The wheel colliders are in place.
![image](https://user-images.githubusercontent.com/44414900/150671378-74218943-4ebe-440a-b99a-968663009017.png)
Could you suggest something?

Thanks ! 🙏😀

**Comment by EricBoiseLGSVL** [comment]
You need to change the code to support VPP wheel colliders, not Unity wheel colliders.  Make sure it is on the default physics layer.  Also, fix all null errors.

**Comment by lemketron** [comment]
> You need to change the code to support VPP wheel colliders, not Unity wheel colliders. Make sure it is on the default physics layer. Also, fix all null errors.

This might help: https://vehiclephysics.com/components/wheel-collider/
